In [23]:
import pandas as pd
import numpy as np

In [24]:
mis_datos = pd.read_csv('..//data/cars.csv')

Here we are asking the program to read an csv file called cars.

Question 1

In [25]:
print("First 5 rows of the dataset:")
display(mis_datos.head())

print("Last 5 rows of the dataset:")
display(mis_datos.tail())

print("Summary statistics of the dataset:")
display(mis_datos.describe())

print("Random samples from the dataset:")
display(mis_datos.sample(n=5))

print("Shape of the dataset:")
display(mis_datos.shape)

print("Column names of the dataset:")
display(mis_datos.columns)

print("Data types of the columns:")
display(mis_datos.dtypes)


First 5 rows of the dataset:


,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
0,Acura,MDX,SUV,Asia,All,36945,33337,3.5,6.0,265.0,17.0,23.0,4451.0,106.0,189.0
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,21761,2.0,4.0,200.0,24.0,31.0,2778.0,101.0,172.0
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,24647,2.4,4.0,200.0,22.0,29.0,3230.0,105.0,183.0
3,Acura,TL 4dr,Sedan,Asia,Front,33195,30299,3.2,6.0,270.0,20.0,28.0,NaN,108.0,186.0
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,39014,3.5,6.0,225.0,18.0,24.0,3880.0,115.0,197.0


Last 5 rows of the dataset:


,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
423,Volvo,C70 LPT convertible 2dr,Sedan,Europe,Front,40565,38203,2.4,5.0,197.0,21.0,28.0,3450.0,105.0,186.0
424,Volvo,C70 HPT convertible 2dr,Sedan,Europe,Front,42565,40083,2.3,5.0,242.0,20.0,26.0,3450.0,105.0,186.0
425,Volvo,S80 T6 4dr,Sedan,Europe,Front,45210,42573,2.9,6.0,268.0,19.0,26.0,3653.0,110.0,190.0
426,Volvo,V40,Wagon,Europe,Front,26135,24641,1.9,4.0,170.0,22.0,29.0,2822.0,101.0,180.0
427,Volvo,XC70,Wagon,Europe,All,35145,33112,2.5,5.0,208.0,20.0,27.0,3823.0,109.0,186.0


Summary statistics of the dataset:


,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
count,428.000000,428.000000,425.000000,425.000000,423.000000,426.000000,426.000000,423.000000,426.000000,427.000000
mean,32774.855140,30014.700935,3.195294,5.811765,215.576832,20.065728,26.840376,3578.229314,108.166667,186.381733
std,19431.716674,17642.117750,1.111932,1.557802,71.978098,5.249574,5.749505,759.034549,8.327582,14.369110
min,10280.000000,9875.000000,1.300000,3.000000,73.000000,10.000000,12.000000,1850.000000,89.000000,143.000000
25%,20334.250000,18866.000000,2.300000,4.000000,165.000000,17.000000,24.000000,3103.000000,103.000000,178.000000
50%,27635.000000,25294.500000,3.000000,6.000000,210.000000,19.000000,26.000000,3476.000000,107.000000,187.000000
75%,39205.000000,35710.250000,3.900000,6.000000,252.500000,21.750000,29.000000,3978.500000,112.000000,194.000000
max,192465.000000,173560.000000,8.300000,12.000000,500.000000,60.000000,66.000000,7190.000000,144.000000,238.000000


Random samples from the dataset:


,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
149,Honda,Civic Hybrid 4dr manual (gas/electric),Hybrid,Asia,Front,20140,18451,1.4,4.0,93.0,46.0,51.0,2732.0,103.0,175.0
333,Porsche,911 Targa coupe 2dr,Sports,Europe,Rear,76765,67128,3.6,6.0,315.0,18.0,26.0,3119.0,93.0,175.0
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,41100,3.5,6.0,225.0,18.0,24.0,3893.0,115.0,197.0
331,Porsche,911 Carrera convertible 2dr (coupe),Sports,Europe,Rear,79165,69229,3.6,6.0,315.0,18.0,26.0,3135.0,93.0,175.0
304,Nissan,Altima S 4dr,Sedan,Asia,Front,19240,18030,2.5,4.0,175.0,21.0,26.0,3039.0,110.0,192.0


Shape of the dataset:


(428, 15)

Column names of the dataset:


Index(['Make', 'Model', 'Type', 'Origin', 'DriveTrain', 'MSRP', 'Invoice',
       'EngineSize', 'Cylinders', 'Horsepower', 'MPG_City', 'MPG_Highway',
       'Weight', 'Wheelbase', 'Length'],
      dtype='object')

Data types of the columns:


Make            object
Model           object
Type            object
Origin          object
DriveTrain      object
MSRP             int64
Invoice          int64
EngineSize     float64
Cylinders      float64
Horsepower     float64
MPG_City       float64
MPG_Highway    float64
Weight         float64
Wheelbase      float64
Length         float64
dtype: object

The dataset cars has 428 observations and 15 variables. It has different types of column data such as objects, float64, and int 64. It also contains both numerical and qualitative data such as the columns model, origin, etc.

Question 2

In [26]:
print("Checking for missing values in the dataset:")
display(mis_datos.isnull().sum())

Checking for missing values in the dataset:


Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     3
Cylinders      3
Horsepower     5
MPG_City       2
MPG_Highway    2
Weight         5
Wheelbase      2
Length         1
dtype: int64

The code tells us that: Engine size, cylinders, horsepower, MPG_City, MPG_ Highway, weight, wheelbase, and length, contain missing values and it gives us the amount of missing values it contains as well.

In [27]:
print("Replace missing values with the mean of the respective columns:")
for column in mis_datos.columns:
    if mis_datos[column].isnull().sum() > 0:
        mean_value = mis_datos[column].mean()
        mis_datos[column].fillna(mean_value, inplace=True)
display(mis_datos.isnull().sum())
        

Replace missing values with the mean of the respective columns:


/tmp/ipykernel_53692/1107201789.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  mis_datos[column].fillna(mean_value, inplace=True)


Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

Here we can see that the previous columns that had missing values now appear to have none.

Question 3

In [19]:
print("Which car manufacturer has the largest number of cars in the dataset? Which car manufacturer has the smallest number of cars in the dataset?")
manufacturer_counts = mis_datos['Make'].value_counts()
largest_manufacturer = manufacturer_counts.idxmax()
smallest_manufacturer = manufacturer_counts.idxmin()
print(f"The car manufacturer with the largest number of cars is: {largest_manufacturer} with {manufacturer_counts.max()} cars.")
print(f"The car manufacturer with the smallest number of cars is: {smallest_manufacturer} with {manufacturer_counts.min()} cars.")

Which car manufacturer has the largest number of cars in the dataset? Which car manufacturer has the smallest number of cars in the dataset?
The car manufacturer with the largest number of cars is: Toyota with 28 cars.
The car manufacturer with the smallest number of cars is: Hummer with 1 cars.


Here we can see that the manufacturer that produces the largest number of vehicles is Toyota with 28 cars and the smallest number is Hummer with 1 car

Question 4

In [20]:
print("How many cars are produced in each region with front-wheel drive, rear-wheel drive, and all-wheel drive?")
drive_counts = (
    mis_datos.groupby(["Origin", "DriveTrain"])
    .size()
    .unstack(fill_value=0))
sum_counts = drive_counts.sum(axis=1)
drive_counts["Total"] = sum_counts

display(drive_counts)

How many cars are produced in each region with front-wheel drive, rear-wheel drive, and all-wheel drive?


DriveTrain,All,Front,Rear,Total
Origin,,,,
Asia,34,99,25,158
Europe,36,37,50,123
USA,22,90,35,147


Here we can see the total of cars produced in each region based on all drives. In Asia the total is 158, in Europe 123, and in USA 147.

Question 5

In [21]:
print("Highest suggested retail price:")
highest_price = mis_datos["MSRP"].max()
display(highest_price)

print("Lowest suggested retail price:")
lowest_price = mis_datos["MSRP"].min()
display(lowest_price)

print("Average city miles per gallon:")
average_city_mpg = mis_datos["MPG_City"].mean()
display(average_city_mpg)

print("Average highway miles per gallon (MPG):")
average_highway_mpg = mis_datos["MPG_Highway"].mean()
display(average_highway_mpg)

Highest suggested retail price:


np.int64(192465)

Lowest suggested retail price:


np.int64(10280)

Average city miles per gallon:


np.float64(20.065727699530516)

Average highway miles per gallon (MPG):


np.float64(26.840375586854464)

The highest suggested retail price was 192465. The lowest suggested retail price was 10280. The average city miles per gallon was 20.065. Lastly, the average highway miles per gallon was 26.84.

Question 6
Select only the following variables: Make, Model, Type, Origin, DriveTrain, MSRP, Weight,
Wheelbase, and Length.
Keep only the observations where DriveTrain is equal to Front, and store this information in
a variable called mis_nuevos_datos.

In [31]:
mis_datos = pd.read_csv('..//data/cars.csv')
print("variables where drivetrain is equal to front")
mis_nuevos_datos = mis_datos.loc[mis_datos["DriveTrain"] == "Front", ["Make","Model","Type", "Origin", "DriveTrain", "MSRP", "Weight", "Wheelbase","Length"]]
display(mis_nuevos_datos)


variables where drivetrain is equal to front


,Make,Model,Type,Origin,DriveTrain,MSRP,Weight,Wheelbase,Length
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,2778.0,101.0,172.0
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,3230.0,105.0,183.0
3,Acura,TL 4dr,Sedan,Asia,Front,33195,NaN,108.0,186.0
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,3880.0,115.0,197.0
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,3893.0,115.0,197.0
...,...,...,...,...,...,...,...,...,...
421,Volvo,S80 2.9 4dr,Sedan,Europe,Front,37730,3576.0,110.0,190.0
423,Volvo,C70 LPT convertible 2dr,Sedan,Europe,Front,40565,3450.0,105.0,186.0
424,Volvo,C70 HPT convertible 2dr,Sedan,Europe,Front,42565,3450.0,105.0,186.0
425,Volvo,S80 T6 4dr,Sedan,Europe,Front,45210,3653.0,110.0,190.0


Here we can see that there are two types which drivetrains are front, they are Sedan and Wagon, and there are a total of 226 observations.

Question 7


In [35]:
print("First observations of mis_nuevos_datos:")
display(mis_nuevos_datos.head())

Weight_kilo = mis_nuevos_datos["Weight"] * 0.453592
mis_nuevos_datos["Weight_kilo"] = Weight_kilo

Wheelbase_cm = mis_nuevos_datos["Wheelbase"] * 0.0254
mis_nuevos_datos["Wheelbase_cm"] = Wheelbase_cm

Length_cm = mis_nuevos_datos["Length"] * 0.0254
mis_nuevos_datos["Length_cm"] = Length_cm

First observations of mis_nuevos_datos:


,Make,Model,Type,Origin,DriveTrain,MSRP,Weight,Wheelbase,Length,Weight_kilo,Wheelbase_meters,Length_meters,Wheelbase_cm,Length_cm
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,2778.0,101.0,172.0,1260.078576,2.5654,4.3688,2.5654,4.3688
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,3230.0,105.0,183.0,1465.102160,2.6670,4.6482,2.6670,4.6482
3,Acura,TL 4dr,Sedan,Asia,Front,33195,NaN,108.0,186.0,NaN,2.7432,4.7244,2.7432,4.7244
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,3880.0,115.0,197.0,1759.936960,2.9210,5.0038,2.9210,5.0038
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,3893.0,115.0,197.0,1765.833656,2.9210,5.0038,2.9210,5.0038


Here the code is calling the first data from mis_nuevos_datos and it's being displayed. After that I made 3 new variables and stored them inside the mis_nuevos_datos variable. 

Question 8

In [37]:
opinion_based_on_suggested_retail_price = mis_nuevos_datos["MSRP"].apply(lambda x: "Expensive" if x > 60000 else "Normal" if 60000 > x > 30000 else "Inexpensive" if x < 30000 else "Unknown")
mis_nuevos_datos["Opinion_based_on_suggested_retail_price"] = opinion_based_on_suggested_retail_price
print("Opinion counts:")
print(mis_nuevos_datos["Opinion_based_on_suggested_retail_price"].value_counts())
display(mis_nuevos_datos)

Opinion counts:
Opinion_based_on_suggested_retail_price
Inexpensive    175
Normal          48
Expensive        3
Name: count, dtype: int64


,Make,Model,Type,Origin,DriveTrain,MSRP,Weight,Wheelbase,Length,Weight_kilo,Wheelbase_meters,Length_meters,Wheelbase_cm,Length_cm,Opinion_based_on_suggested_retail_price
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,2778.0,101.0,172.0,1260.078576,2.5654,4.3688,2.5654,4.3688,Inexpensive
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,3230.0,105.0,183.0,1465.102160,2.6670,4.6482,2.6670,4.6482,Inexpensive
3,Acura,TL 4dr,Sedan,Asia,Front,33195,NaN,108.0,186.0,NaN,2.7432,4.7244,2.7432,4.7244,Normal
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,3880.0,115.0,197.0,1759.936960,2.9210,5.0038,2.9210,5.0038,Normal
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,3893.0,115.0,197.0,1765.833656,2.9210,5.0038,2.9210,5.0038,Normal
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
421,Volvo,S80 2.9 4dr,Sedan,Europe,Front,37730,3576.0,110.0,190.0,1622.044992,2.7940,4.8260,2.7940,4.8260,Normal
423,Volvo,C70 LPT convertible 2dr,Sedan,Europe,Front,40565,3450.0,105.0,186.0,1564.892400,2.6670,4.7244,2.6670,4.7244,Normal
424,Volvo,C70 HPT convertible 2dr,Sedan,Europe,Front,42565,3450.0,105.0,186.0,1564.892400,2.6670,4.7244,2.6670,4.7244,Normal
425,Volvo,S80 T6 4dr,Sedan,Europe,Front,45210,3653.0,110.0,190.0,1656.971576,2.7940,4.8260,2.7940,4.8260,Normal


In this last chunk we are making an if statement. If the price is greater than 60,000 then it is expensive, if the car is between the prices 30,000 and 60,000 the price is normal, lastly if the price is below 30,000 then the car is inexpensive. Then we displayed the whole new data, the last column shows the price classification. At the very top we can see the number of each opinion retial price, we have 175 inexpensive, 48 normal, and 3 expensive.